Block Library Implementation, Prefix Moving

svouse opened the pull request:
**Summary:** Implements all non-export functions in block_library.js, moves definition of prefix constants to app_controller.js.

**Goal:** Maintaining functionality, fill in BlockLibrary functions (with resulting changes to other classes).

**Specifically Includes:**
- (based on changes in #121 ) distributing prefixes through AppController (due to the fact that resources now have access to their prefix, this will result in cleaning of API, mostly in NavigationTree; however, this will come with the post-model PRs visual update)
-  Implementation of all functions in the BlockLibrary

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/123)
<!-- Reviewable:end -->

svouse committed:
adding prefix to Resource

svouse committed:
changing ownership of prefixes

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into block_library

svouse committed:
updating project passing prefixes

svouse committed:
updates to AppController

svouse committed:
param updates

svouse committed:
param updates

svouse committed:
rewiring completed

svouse committed:
fixed merge conflicts

svouse committed:
resolved differences in app_controller.js

svouse committed:
fixed throws in set

svouse committed:
no crashes

svouse committed:
fixed function calls

rachel-fenichel commented:
let's chat offline about the prefix list.

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/controller/project_controller.js, line 43 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX0umhbdjkMr_hsIMt:-KpX0umhbdjkMr_hsIMu:b4ml81z) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L43)):*
> ```JavaScript
>    * @param {!NavigationTree} tree The tree which represents that project, to be updated
>    *     by the controller alongside it.
>    * @param {Array.<string>} prefixList List of prefixes for all Resource classes.
> ```

Non-null: `!Array.<string>`

---

*[src/model/block_definition.js, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1ICEqisP9_QVZ22g:-KpX1ICEqisP9_QVZ22h:b-665v5r) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_definition.js#L34)):*
> ```JavaScript
>    * BlockDefinition Class.
>    * @constructor
>    * @param {string} type The name of the block.
> ```

Add param: prefix

---

*[src/model/block_library.js, line 111 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1SYxl_Kx3OAawfwC:-KpX1SYxl_Kx3OAawfwD:bqp9dkc) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_library.js#L111)):*
> ```JavaScript
>    */
>   clear() {
>     this.blocks = {};
> ```

`new Object(null)`

---

*[src/model/resource.js, line 35 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1uqbntgQ-GLTcx0p:-KpX1uqcy-UZApRSIQ56:b10jhun) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/resource.js#L35)):*
> ```JavaScript
>    * Resource Class.
>    * @param {string} resourceName The name for the resource.
>    * @param {string} prefix The prefix used to identify the Resource in its Json.
> ```

JSON is all caps, here and below.

---

*[src/model/resource.js, line 87 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX2ZfU6_wqX50YO8Ga:-KpX2ZfU6_wqX50YO8Gb:b-cv2pm4) ([raw file](https://github.com/google/blockly-devtools/blob/07d4241e05892372e708ebdce3f1fe129f812c48/src/model/resource.js#L87)):*
> ```JavaScript
>    * @return {!Object} The tree-specific JSON representation of the array.
>    */
>   makeTreeJsonFromArray(nameArray) {
> ```

Where did this go?

---

*[src/view/navigation_tree.js, line 56 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX2qyAxV8tpVGoqwzc:-KpX2qyAxV8tpVGoqwzd:b-z4wz6n) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/view/navigation_tree.js#L56)):*
> ```JavaScript
>     this.makeTree();
> 
>     this.prefixList = prefixList;
> ```

Annotate this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123#-:-KpX36abFWXTtEe4X9Nx:b-3wuw45)*
<!-- Sent from Reviewable.io -->

chynu commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/controller/app_controller.js, line 54 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX4t3m0Hk3BemF7SRH:-KpX4t3m0Hk3BemF7SRI:bcbe24q) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/app_controller.js#L54)):*
> ```JavaScript
> const PROJECT_PREFIX = "Project";
> const BLOCK_PREFIX = 'Block';
> const LIBRARY_PREFIX = "BlockLibrary";
> ```

Nit: single quotes?

---

*[src/controller/app_controller.js, line 63 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX4QkqeVbln27fO1Fd:-KpX4Qkr-JMPUN3nW4Eb:b-ambkvk) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/app_controller.js#L63)):*
> ```JavaScript
> const prefixList =
>   [PROJECT_PREFIX, BLOCK_PREFIX, LIBRARY_PREFIX, TOOLBOX_PREFIX,
>     WORKSPACE_CONFIG_PREFIX, WORKSPACE_CONTENTS_PREFIX];
> ```

It may be easier if you created an Object for this instead:
```JavaScript
const PREFIX = {
  PROJECT: 'Project',
  BLOCK: 'Block',
  LIBRARY: 'BlockLibrary'
  // and so on
}
```

That way, you can iterate through them all at once by calling `PREFIX`, or easily refer to them as `PREFIX.PROJECT`, etc.

---

*[src/controller/project_controller.js, line 46 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX5EgVKRyqcTDGfK_h:-KpX5EgVKRyqcTDGfK_i:b-7idzml) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L46)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(project, tree, prefixList) {
> ```

I think that since `prefixList` was declared at the global level, referring to it within the code as just `prefixList` without associating it with any object will still work. Either way, `prefixList` is a constant and doesn't need to be tied with objects.

If anything, you can probably attach it at the bottom of `AppController` as static fields to the class:

```JavaScript
class AppController {
  // code blah
}

AppController.PREFIX_LIST = [];
```

That way, anywhere in the code, you don't need a field to reference `AppController.PREFIX_LIST`.

---

*[src/controller/project_controller.js, line 103 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX60VdWPUYtrQ2KXmy:-KpX60VdWPUYtrQ2KXmz:bby1kgf) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L103)):*
> ```JavaScript
>   createBlockDefinition(blockType, libraryName) {
>     //TODO #105: check for valid name, throw error upon conflict
>     const block = new BlockDefinition(blockType, this.prefixList[1]);
> ```

I think using an object/dictionary will make this much more readable.

---

*[src/model/block_library.js, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX6yCapL4lybx17Txt:-KpX6yCapL4lybx17Txu:bpuu5pe) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_library.js#L53)):*
> ```JavaScript
>      * @type {!Object<string, !BlockDefinition>}
>      */
>     this.blocks = {};
> ```

This isn't new code but for the purposes of keeping block names alphabetized, I think that `this.blocks` should be an array of tuple-arrays. We can use the `sort()` function with a defined `compare()` function (I believe JS allows passing in a compare function as a parameter that will let you define what goes 'before' the other). That way, we can keep the list of blocks ordered alphabetically, and consistently so. (I think JS objects don't necessarily keep the keys in the order you declare them in?)

Also, perhaps we should have a `sortBlocks()` function within this class that we can call on `this` so that we can run it any time we add a new block.

Another alternative is keeping it as a map and sorting right before we return the list of block types.

We can talk about this more offline since this affects most of the model code (as we will likely alphabetize the elements in the sets as well). Dictionaries might make alphabetizing difficult.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR fixes, changed prefixList to dictionary

svouse committed:
rewired

svouse committed:
documentation updates

svouse commented:
Review status: 4 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/controller/app_controller.js, line 54 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX4t3m0Hk3BemF7SRH:-KpXBvU7f58-OS6fuI8P:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/app_controller.js#L54)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Nit: single quotes?
</blockquote></details>

Done.

---

*[src/controller/app_controller.js, line 63 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX4QkqeVbln27fO1Fd:-KpXC1ZrGhESPvUaXRpx:b-j09gic) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/app_controller.js#L63)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

It may be easier if you created an Object for this instead:
```JavaScript
const PREFIX = {
  PROJECT: 'Project',
  BLOCK: 'Block',
  LIBRARY: 'BlockLibrary'
  // and so on
}
```

That way, you can iterate through them all at once by calling `PREFIX`, or easily refer to them as `PREFIX.PROJECT`, etc.
</blockquote></details>

I dig it; done//Thanks!

---

*[src/controller/project_controller.js, line 43 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX0umhbdjkMr_hsIMt:-KpXMBuGMJd96oP56fL2:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L43)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Non-null: `!Array.<string>`
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 46 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX5EgVKRyqcTDGfK_h:-KpXMOgP6tm5dz4asS-C:bl4k4fa) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L46)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think that since `prefixList` was declared at the global level, referring to it within the code as just `prefixList` without associating it with any object will still work. Either way, `prefixList` is a constant and doesn't need to be tied with objects.

If anything, you can probably attach it at the bottom of `AppController` as static fields to the class:

```JavaScript
class AppController {
  // code blah
}

AppController.PREFIX_LIST = [];
```

That way, anywhere in the code, you don't need a field to reference `AppController.PREFIX_LIST`.
</blockquote></details>

I dig it (also changed to dictionary)

---

*[src/controller/project_controller.js, line 103 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX60VdWPUYtrQ2KXmy:-KpXCETwVUlfSAISa-Qm:b-vmop99) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/controller/project_controller.js#L103)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think using an object/dictionary will make this much more readable.
</blockquote></details>

Done//Correct

---

*[src/model/block_definition.js, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1ICEqisP9_QVZ22g:-KpXMc4Ux0vRB9rkn4_k:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_definition.js#L34)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Add param: prefix
</blockquote></details>

Done.

---

*[src/model/block_library.js, line 53 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX6yCapL4lybx17Txt:-KpXQ75KLm6zt27-k4zj:bnvijdz) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_library.js#L53)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

This isn't new code but for the purposes of keeping block names alphabetized, I think that `this.blocks` should be an array of tuple-arrays. We can use the `sort()` function with a defined `compare()` function (I believe JS allows passing in a compare function as a parameter that will let you define what goes 'before' the other). That way, we can keep the list of blocks ordered alphabetically, and consistently so. (I think JS objects don't necessarily keep the keys in the order you declare them in?)

Also, perhaps we should have a `sortBlocks()` function within this class that we can call on `this` so that we can run it any time we add a new block.

Another alternative is keeping it as a map and sorting right before we return the list of block types.

We can talk about this more offline since this affects most of the model code (as we will likely alphabetize the elements in the sets as well). Dictionaries might make alphabetizing difficult.
</blockquote></details>

I dig it; made an issue for alphabetizing, and invited you to a meeting tomorrow (although high key if the time doesn't work we can talk about it whenever)

---

*[src/model/block_library.js, line 111 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1SYxl_Kx3OAawfwC:-KpXCGwyLk8-YIR11lEM:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/block_library.js#L111)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

`new Object(null)`
</blockquote></details>

Done.

---

*[src/model/resource.js, line 35 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX1uqbntgQ-GLTcx0p:-KpXEE34JqG5ZLHwaiMJ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/model/resource.js#L35)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

JSON is all caps, here and below.
</blockquote></details>

Done.

---

*[src/model/resource.js, line 87 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX2ZfU6_wqX50YO8Ga:-KpXRM1ZbDI0EgVnytl0:byyovk2) ([raw file](https://github.com/google/blockly-devtools/blob/07d4241e05892372e708ebdce3f1fe129f812c48/src/model/resource.js#L87)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Where did this go?
</blockquote></details>

I'd removed it initially thinking that the classes which extend resource would need to define their own getJson methods, but thought about it and now it's back.

---

*[src/view/navigation_tree.js, line 56 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX2qyAxV8tpVGoqwzc:-KpXRXCQVYvZF8LbCnUn:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/view/navigation_tree.js#L56)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Annotate this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
added an asterisk above comment for prefix dictionary, sinc eit now refers to one object directly below the notation

chynu commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/controller/project_controller.js, line 63 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpXc1TGQBTJ5Z9N43xL:-KpXc1TGQBTJ5Z9N43xM:b-4443rj) ([raw file](https://github.com/google/blockly-devtools/blob/956156d828fed4e71a3b9d2866e05029b6806873/src/controller/project_controller.js#L63)):*
> ```JavaScript
>      * @type {Object}
>      */
>     this.prefixes = prefixes;
> ```

I think you can just refer to the prefix object without saving it into the object definition itself (i.e. whenever you need `PREFIXES` you can just refer to it anywhere since it has been declared globally within the application).

This applies to everywhere else in this PR that you pass `PREFIXES` in as a parameter. That way, there is less dependency on a constant that shouldn't change based on what object is loaded, anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
no longer passing prefixes dictionary

svouse commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/controller/project_controller.js, line 63 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpXc1TGQBTJ5Z9N43xL:-Kpa-x6abJ7-WAoFwz6P:b-vg9oq5) ([raw file](https://github.com/google/blockly-devtools/blob/956156d828fed4e71a3b9d2866e05029b6806873/src/controller/project_controller.js#L63)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think you can just refer to the prefix object without saving it into the object definition itself (i.e. whenever you need `PREFIXES` you can just refer to it anywhere since it has been declared globally within the application).

This applies to everywhere else in this PR that you pass `PREFIXES` in as a parameter. That way, there is less dependency on a constant that shouldn't change based on what object is loaded, anyway.
</blockquote></details>

I dig it; thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

chynu commented:
Review status: 4 of 12 files reviewed at latest revision, 17 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaH_eAeQI7xEPy0iEg:-KpaH_eAeQI7xEPy0iEh:b-5pdr5s) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/controller/app_controller.js#L50)):*
> ```JavaScript
>  * Dictionary containing the prefixes for all node types. Given with the
>  *     assumption that the name of each object in a project is unique across
>  *     that project.
> ```

Nit: add `@type`? Although I guess this doesn't exactly fall under a specific class.

---

*[src/model/block_definition.js, line 37 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHpKBgYsvjJ5blPS_:-KpaHpKBgYsvjJ5blPSa:b-jh38wx) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_definition.js#L37)):*
> ```JavaScript
>    * @param {!string} prefix The prefix for identifying blocks.
>    */
>   constructor(type, prefix) {
> ```

Remove `prefix` param? You can probably just refer to `PREFIXES.BLOCK` whenever you need it within this class. The `prefix` param will be the same for every instance of `BlockDefinition` anyway, so I don't think it needs to be part of the constructor (which kind of implies that they can vary between instances of the class). I think adding constants into a constructor that doesn't vary between instances just adds more params to worry about.

Another thing is making a static field within the class, either by just doing

```JavaScript
static get PREFIX() {
  return PREFIXES.BLOCK;
}
```

or

```JavaScript
class BlockDefinition extends Resource {
  // code
}

BlockDefinition.PREFIX = PREFIXES.BLOCK;
```
(The same reasoning applies to the other model objects/constructors.)

---

*[src/model/block_library.js, line 41 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHtVZLTG2Xae0spaD:-KpaHtVZLTG2Xae0spaE:b8l8a19) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L41)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(libraryName, prefix) {
> ```

Remove `prefix` param?

---

*[src/model/block_library.js, line 47 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHv83en5CF32X0Wl5:-KpaHv83en5CF32X0Wl6:b8l8a19) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L47)):*
> ```JavaScript
>      * References: N/A
>      */
>     super(libraryName, prefix);
> ```

Remove `prefix` param?

---

*[src/model/block_library.js, line 111 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHyIB6DqG1S5FUy2_:-KpaHyIB6DqG1S5FUy2a:bue1o61) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L111)):*
> ```JavaScript
>    */
>   clear() {
>     this.blocks = new Object(null);
> ```

I think per our discussion yesterday, it should be `Object.create(null)`.

---

*[src/model/block_library_set.js, line 40 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaI8spxmCy8ebK3L7U:-KpaI8spxmCy8ebK3L7V:b8l8a19) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library_set.js#L40)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(librarySetName, projectName, prefix) {
> ```

Remove `prefix` param?

---

*[src/model/block_library_set.js, line 101 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIAuITPpPSgefvfA5:-KpaIAuITPpPSgefvfA6:b-tn8sln) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library_set.js#L101)):*
> ```JavaScript
>   getJson() {
>     const librarySetJson = {
>       'id': this.prefix,
> ```

Replace with `PREFIX.LIBRARY`?

---

*[src/model/resource.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIN0r9NbeeByaUEFh:-KpaIN0r9NbeeByaUEFi:b8l8a19) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource.js#L38)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(resourceName, prefix) {
> ```

Remove `prefix` param?

---

*[src/model/resource.js, line 82 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIUSoAguLEGV4jo47:-KpaIUSoAguLEGV4jo48:b-rduc7m) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource.js#L82)):*
> ```JavaScript
>   getJson() {
>     const resourceJson = {
>       'id': this.prefix,
> ```

Remove `this.prefix` reference, and define `'id'` field in each subclass?

---

*[src/model/resource_set.js, line 42 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIfPhSl7z9mM8vJXn:-KpaIfPhSl7z9mM8vJXo:be03nah) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource_set.js#L42)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(resourceSetName, projectName, prefix,) {
> ```

Remove prefix param?

---

*[src/view/navigation_tree.js, line 182 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaKrB6Os-LepWJIxI0:-KpaKrB6Os-LepWJIxI1:bdvm54u) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/view/navigation_tree.js#L182)):*
> ```JavaScript
>    */
>   deleteBlockNode(blockType) {
>     this.deleteComponentNode(prefixList[1], blockType);
> ```

Change reference to `PREFIXES`? Here and below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 4 of 8 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[src/model/block_library.js, line 111 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHyIB6DqG1S5FUy2_:-Kpajz5kbdwt29szobX6:b32v57) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L111)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think per our discussion yesterday, it should be `Object.create(null)`.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR responses

svouse committed:
PR responses

svouse commented:
Review status: 1 of 11 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[src/model/block_library.js, line 41 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHtVZLTG2Xae0spaD:-KpazC5JByKeJaxsMJcL:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L41)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `prefix` param?
</blockquote></details>

Done.

---

*[src/model/block_library.js, line 47 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHv83en5CF32X0Wl5:-KpazCZtcIjB7yxsTAD5:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L47)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `prefix` param?
</blockquote></details>

Done.

---

*[src/model/block_library.js, line 111 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHyIB6DqG1S5FUy2_:-Kpaz5I-uIDX1SHkK5M2:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library.js#L111)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[src/model/block_library_set.js, line 40 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaI8spxmCy8ebK3L7U:-Kpaz0iLCQnPfJVrXwfY:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library_set.js#L40)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `prefix` param?
</blockquote></details>

Done.

---

*[src/model/block_library_set.js, line 101 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIAuITPpPSgefvfA5:-Kpaz12Ifq7DgJCXm0qL:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_library_set.js#L101)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Replace with `PREFIX.LIBRARY`?
</blockquote></details>

Done.

---

*[src/model/resource.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIN0r9NbeeByaUEFh:-Kpaz-t1HLk7lcgC2jOZ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource.js#L38)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `prefix` param?
</blockquote></details>

Done.

---

*[src/model/resource.js, line 82 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIUSoAguLEGV4jo47:-Kpaz-DL0Ag9vu3BiaZ-:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource.js#L82)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `this.prefix` reference, and define `'id'` field in each subclass?
</blockquote></details>

Done.

---

*[src/model/resource_set.js, line 42 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaIfPhSl7z9mM8vJXn:-KpayzW9U-ufsc65GDsu:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/resource_set.js#L42)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove prefix param?
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 182 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaKrB6Os-LepWJIxI0:-Kpayy-ELg83HnRAoa5U:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/view/navigation_tree.js#L182)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Change reference to `PREFIXES`? Here and below
</blockquote></details>

Done.

---

*[src/model/block_definition.js, line 37 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaHpKBgYsvjJ5blPS_:-KpaLfDZSLpqTboVAvmX:b-vrlytz) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/model/block_definition.js#L37)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Remove `prefix` param? You can probably just refer to `PREFIXES.BLOCK` whenever you need it within this class. The `prefix` param will be the same for every instance of `BlockDefinition` anyway, so I don't think it needs to be part of the constructor (which kind of implies that they can vary between instances of the class). I think adding constants into a constructor that doesn't vary between instances just adds more params to worry about.

Another thing is making a static field within the class, either by just doing

```JavaScript
static get PREFIX() {
  return PREFIXES.BLOCK;
}
```

or

```JavaScript
class BlockDefinition extends Resource {
  // code
}

BlockDefinition.PREFIX = PREFIXES.BLOCK;
```
(The same reasoning applies to the other model objects/constructors.)
</blockquote></details>

I dig it;got rid of all prefix params

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 1 of 11 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaH_eAeQI7xEPy0iEg:-Kpb-bxBdxw1aoiei2hU:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Nit: add `@type`? Although I guess this doesn't exactly fall under a specific class.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

chynu commented:
Review status: 1 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaH_eAeQI7xEPy0iEg:-Kpb2vtEhrEHYrR4KB3b:b-i39erm) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Done.
</blockquote></details>

fun fact I think you can do `{!Object.<string, string>}` as well, though this is A Very Nitpicky Nit haha /__\

---

*[src/controller/project_controller.js, line 96 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3YqYvR1fXGomCaRl:-Kpb3YqYvR1fXGomCaRm:b4uy845) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L96)):*
> ```JavaScript
>   createBlockDefinition(blockType, libraryName) {
>     //TODO #105: check for valid name, throw error upon conflict
>     const block = new BlockDefinition(blockType, PREFIXES.BLOCK);
> ```

I think you forgot to remove the prefix param here.

---

*[src/controller/project_controller.js, line 125 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3bRuVBePxA5CsqFB:-Kpb3bRuVBePxA5CsqFC:b-et929n) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L125)):*
> ```JavaScript
>   createToolbox(toolboxName) {
>     //TODO #105: check for valid name, throw error upon conflict
>     const toolbox = new Toolbox(toolboxName, PREFIXES.TOOLBOX);
> ```

We no longer need this in the constructor, I think?

---

*[src/controller/project_controller.js, line 141 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3ddtyf_M3AUgso4A:-Kpb3ddtyf_M3AUgso4B:bs8aex0) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L141)):*
> ```JavaScript
>     //TODO #105: check for valid name, throw error upon conflict
>     const workspaceContents = new WorkspaceContents(workspaceContentsName,
>       PREFIXES.WORKSPACE_CONTENTS);
> ```

Prefix param no longer necessary?

---

*[src/controller/project_controller.js, line 158 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3jAKnCwji9Rx_flH:-Kpb3jAKnCwji9Rx_flI:b-f99z15) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L158)):*
> ```JavaScript
>     //TODO #105: check for valid name, throw error upon conflict
>     const workspaceConfig = new WorkspaceConfiguration(workspaceConfigName);
>     this.addWorkspaceConfiguration(workspaceConfig, PREFIXES.WORKSPACE_CONFIG);
> ```

Do we need this parameter if the constant will always be the same whenever you call `addWorkspaceConfiguration()`?

---

*[src/model/toolbox_set.js, line 40 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3JZJbhb5epvKnPr4:-Kpb3JZJbhb5epvKnPr5:b-p6jlcv) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/model/toolbox_set.js#L40)):*
> ```JavaScript
>    * @constructor
>    */
>   constructor(toolboxSetName, projectName, prefix) {
> ```

I think you forgot to remove the prefix param here and below in line 41.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR responses

svouse commented:
Review status: 0 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpaH_eAeQI7xEPy0iEg:-Kpb4dJfLbBaZ3J889Ik:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/40976ff3c5559d0c5625ce552cd9c499d7abaa60/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

fun fact I think you can do `{!Object.<string, string>}` as well, though this is A Very Nitpicky Nit haha /__\
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 96 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3YqYvR1fXGomCaRl:-Kpb4e2URXvaYulXMgpU:bb73gek) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L96)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think you forgot to remove the prefix param here.
</blockquote></details>

Rip 

---

*[src/controller/project_controller.js, line 125 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3bRuVBePxA5CsqFB:-Kpb4hZ1vJk1UMfPj0uV:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L125)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

We no longer need this in the constructor, I think?
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 141 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3ddtyf_M3AUgso4A:-Kpb4i9nBpASRoU_W8Tj:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L141)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Prefix param no longer necessary?
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 158 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3jAKnCwji9Rx_flH:-Kpb4icPRl_BuYLXnv7q:b-ss7tb5) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/controller/project_controller.js#L158)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Do we need this parameter if the constant will always be the same whenever you call `addWorkspaceConfiguration()`?
</blockquote></details>

Nope. Done.

---

*[src/model/toolbox_set.js, line 40 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb3JZJbhb5epvKnPr4:-Kpb6CvDT6fE45LwW_DC:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/9eba69c05dbb15336e9214f128e43080c90daa35/src/model/toolbox_set.js#L40)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think you forgot to remove the prefix param here and below in line 41.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

chynu commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> ye

---

Review status: 0 of 11 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123#-:-Kpb8Q4WxmpLgovQtyh7:bks4n97)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 6 of 11 files at r4, 2 of 2 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb96e4bbJ3h9FLt4ku:-Kpb96e4bbJ3h9FLt4kv:b-y9967p) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/app_controller.js#L50)):*
> ```JavaScript
>  * Dictionary containing the prefixes for all node types. Given with the
>  *     assumption that the name of each object in a project is unique across
>  *     that project.
> ```

Does this mean a toolbox cannot have the same name as a workspace?

---

*[src/controller/app_controller.js, line 59 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpbBJXvnzNiFdjXyv2Q:-KpbBJXvnzNiFdjXyv2R:b5w9r8l) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/app_controller.js#L59)):*
> ```JavaScript
>   TOOLBOX : 'Toolbox',
>   WORKSPACE_CONTENTS : 'WorkspaceContents',
>   WORKSPACE_CONFIG : 'WorkspaceConfiguration'
> ```

If we want to reinforce immutability, this should be a class with static getters.  The `const` protects `PREFIXES`, but not any of the member references.

---

*[src/controller/project_controller.js, line 112 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb9LhaCE9akbpRHiLV:-Kpb9LhaCE9akbpRHiLW:b-kdyvna) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/project_controller.js#L112)):*
> ```JavaScript
>     //TODO #105: check for valid name, throw error upon conflict
>     const blockLibrary = new BlockLibrary(blockLibraryName);
>     this.addBlockLibrary(blockLibrary);
> ```

This uses `addBlockLibrary(..)`. Below, I see `getLibrary(..)`. Whether `Library` or `BlockLibrary`, it should be consistent.

---

*[src/model/toolbox_set.js, line 35 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpbAVPJVmO7DaRWovEC:-KpbAVPJVmO7DaRWovED:b-bk8fg5) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/model/toolbox_set.js#L35)):*
> ```JavaScript
>    * ToolboxSet Class.
>    * @param {!string} toolboxSetName The name for the toolbox set.
>    * @param {!string} projectName The name of the project the set belongs to.
> ```

`string`, `boolean`, and `number` types are never null; no need for `!`.  (I always find this Closure notation strange.)

---

*[src/view/navigation_tree.js, line 56 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpX2qyAxV8tpVGoqwzc:-KpbBAWekKS1HoFRhswa:bss4a2b) ([raw file](https://github.com/google/blockly-devtools/blob/27624787ed9eea528bffb5e6ceb6cb0949b02909/src/view/navigation_tree.js#L56)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 7 of 11 files at r4, 2 of 2 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
made prefixes a class, made BlockLibrary method names consistent

svouse committed:
updated getLibrary

svouse commented:
Review status: 7 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb96e4bbJ3h9FLt4ku:-KpbHGAFpFtZ13pbV1EK:b2nhtan) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Does this mean a toolbox cannot have the same name as a workspace?
</blockquote></details>

Yup

---

*[src/controller/app_controller.js, line 59 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpbBJXvnzNiFdjXyv2Q:-KpbQe3Ng7UChiDHNqV0:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/app_controller.js#L59)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

If we want to reinforce immutability, this should be a class with static getters.  The `const` protects `PREFIXES`, but not any of the member references.
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 112 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb9LhaCE9akbpRHiLV:-KpbLLCkEXC0dgSku1DB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/project_controller.js#L112)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This uses `addBlockLibrary(..)`. Below, I see `getLibrary(..)`. Whether `Library` or `BlockLibrary`, it should be consistent.
</blockquote></details>

Done.

---

*[src/model/toolbox_set.js, line 35 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpbAVPJVmO7DaRWovEC:-KpbLPy4xD8zUq2LMEyZ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/model/toolbox_set.js#L35)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

`string`, `boolean`, and `number` types are never null; no need for `!`.  (I always find this Closure notation strange.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 7 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpb96e4bbJ3h9FLt4ku:-KppisxY7MQ2TaMfsFbq:bcnewt5) ([raw file](https://github.com/google/blockly-devtools/blob/80dbb139eee4f91a61d08309aae9e1dfa08eb66e/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Yup
</blockquote></details>

Hmmm..... Not sure how I feel about this, but no need to change.

---

*[src/controller/app_controller.js, line 50 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KppgUqjy_u1ktqOJ2xk:-KppgUqjy_u1ktqOJ2xl:b-kcg5km) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/controller/app_controller.js#L50)):*
> ```JavaScript
>  * Class containing static getters for the prefixes of all node types. Given
>  *     with the assumption that the name of each object in a project is unique
>  *     across that project.
> ```

All lines of a documentation paragraph should start at the same column.

---

*[src/model/block_library.js, line 87 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpphi-3krWn7JVPfdB_:-Kpphi-3krWn7JVPfdBa:baz7wor) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L87)):*
> ```JavaScript
>    * @return {Object} The JSON that represents the block type or null.
>    */
>   getBlockJson(blockType) {
> ```

Replace this method with `getBlockArrayJson()` that returns all the blocks.

---

*[src/model/block_library.js, line 148 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpphBGsItGFL0RaQy2j:-KpphBGsItGFL0RaQy2k:b-nyaoi9) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L148)):*
> ```JavaScript
>    */
>   has(blockType) {
>     return this.getBlockTypes().includes(blockType);
> ```

`return !!this.blocks[blockType]`?

---

*[src/model/block_library.js, line 167 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpphTk52tT3s1ktxr4P:-KpphTk52tT3s1ktxr4Q:b-hfoogq) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L167)):*
> ```JavaScript
>   getJson() {
>     const libraryJson = $.extend(true, super.getJson(),
>       {'id': PREFIXES.LIBRARY, 'children': this.getBlockJson()});
> ```

`getBlockJson()` takes a `blockType`.  I recommend replacing `this.getBlockJson()` with a new method `this.getBlockArrayJson()`.

---

*[src/view/navigation_tree.js, line 132 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KppiRjS1rwrzC-ydQY0:-KppiRjS1rwrzC-ydQY1:bq575wg) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/view/navigation_tree.js#L132)):*
> ```JavaScript
>   addToolboxNode(toolboxName) {
>     this.addComponentNode(PREFIXES.TOOLBOX, toolboxName,
>       PREFIXES.TOOLBOX);
> ```

Either unwrap this line, or double indent the continuations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 7 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KppgUqjy_u1ktqOJ2xk:-KppjKs0O99jhCHu572B:b-kluzpi) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

All lines of a documentation paragraph should start at the same column.
</blockquote></details>

The "indent +4" rule applies for the description of a parameter or return value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR responses

svouse committed:
changed 'bocks' to 'blocks'

svouse commented:
Review status: 8 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/app_controller.js, line 50 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KppgUqjy_u1ktqOJ2xk:-Kppjr0NNa-fVT7I5m2q:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/controller/app_controller.js#L50)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

The "indent +4" rule applies for the description of a parameter or return value.
</blockquote></details>

Done.

---

*[src/model/block_library.js, line 87 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-Kpphi-3krWn7JVPfdB_:-Kpq7jLvtv8qapikeDZo:bd22w53) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L87)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Replace this method with `getBlockArrayJson()` that returns all the blocks.
</blockquote></details>

Done; added a TODO #130 which talks about what this will look like, and removed getBlockJsonMap (because the json object can be used in the same way the map would have)

---

*[src/model/block_library.js, line 148 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpphBGsItGFL0RaQy2j:-KppnI2gHXKsK3KDeNdy:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L148)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

`return !!this.blocks[blockType]`?
</blockquote></details>

Done.

---

*[src/model/block_library.js, line 167 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpphTk52tT3s1ktxr4P:-KppnP30LF4X-d8SOaGs:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/model/block_library.js#L167)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

`getBlockJson()` takes a `blockType`.  I recommend replacing `this.getBlockJson()` with a new method `this.getBlockArrayJson()`.
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 132 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KppiRjS1rwrzC-ydQY0:-KppnPpm7m5qYamgb2Nc:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/acca2a6ed5f9a6c18a0f313a6f8ff787a7732b17/src/view/navigation_tree.js#L132)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Either unwrap this line, or double indent the continuations.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 1 of 4 files at r7, 2 of 3 files at r8.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/model/block_library.js, line 87 at r8](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpqBIt-bGVrUfeV-FdE:-KpqBIt1Wmv_JzM7iFyz:b-oosuc4) ([raw file](https://github.com/google/blockly-devtools/blob/51806d6e67e0209223d6ab4521fc0ce513263a51/src/model/block_library.js#L87)):*
> ```JavaScript
>    */
>   getBlockArrayJson() {
>     let blockArrayJson = {};
> ```

I think we want an array to insure we retain ordering.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR responses

svouse commented:
Review status: 9 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/model/block_library.js, line 87 at r8](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpqBIt-bGVrUfeV-FdE:-KpqCXLEZq-yEdGwFLpY:b-bu91xm) ([raw file](https://github.com/google/blockly-devtools/blob/51806d6e67e0209223d6ab4521fc0ce513263a51/src/model/block_library.js#L87)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I think we want an array to insure we retain ordering.
</blockquote></details>

Updated

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 9 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123#-:-KpqDJjwCb-e96YMIB4m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 9 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/model/block_library.js, line 136 at r9](https://reviewable.io:443/reviews/google/blockly-devtools/123#-KpqD6fOs0gjWKjmGAbF:-KpqD6fOs0gjWKjmGAbG:beliznc) ([raw file](https://github.com/google/blockly-devtools/blob/f2ae0b104ba70a28051400176e2c5ff8714853bd/src/model/block_library.js#L136)):*
> ```JavaScript
>    * @return {!Object<string, Object>} Map of block type to corresponding JSON.
>    */
>   getBlockJsonMap() {
> ```

This doesn't get JSON.  This gets XML.
If it is not used, remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123)*
<!-- Sent from Reviewable.io -->

svouse committed:
removed getBlockJsonMap()

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 9 of 11 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/123#-:-KpqEQlOrXW3VwwNYXff:bnfp4nl)*
<!-- Sent from Reviewable.io -->
